File: .ja/workflows/audit/snapshot.py
#!/usr/bin/env python3
"""Usage: snapshot.py   (audit payload JSON on stdin)

audit の 1 実行を $HOME/.claude/workspace/history/ に記録し、直近の prior
snapshot に対する resolved/new/carried の delta を計算する。

stdin:  JSON {scope, focus, pre_flight, raw_findings[], findings[], skipped[]}
        各 raw_findings entry は最低限 {file, message} を持つ。
stdout: 書き込んだ JSON record のパス。
exit 0 は成功。exit 1 は payload が parse 不能 (何も書かない)。

record に追加される解決済みフィールド (シェル / prior snapshot 由来):
  branch        git rev-parse --abbrev-ref HEAD ("unknown" にフォールバック)
  generated_at  UTC ISO-8601
  delta         直近の audit-*.json に対する {resolved, new, carried} カウント。
                (file, message) で照合。初回は全て 0 + note "first run"。
"""

import glob
import json
import os
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import NoReturn

HISTORY_DIR = Path(os.path.expanduser("~")) / ".claude" / "workspace" / "history"


def fail(message) -> NoReturn:
    print(f"Error: {message}", file=sys.stderr)
    sys.exit(1)


def finding_key(f):
    return (f.get("file"), f.get("message"))


def compute_delta(current_raw, prior_raw):
    """resolved = prior のみ、new = current のみ、carried = 両方に存在。

    prior snapshot が無いとき prior_raw は None。その場合は "first run" note を
    記録する。
    """
    if prior_raw is None:
        return {"resolved": 0, "new": 0, "carried": 0, "note": "first run"}
    cur = {finding_key(f) for f in current_raw}
    prior = {finding_key(f) for f in prior_raw}
    return {
        "resolved": len(prior - cur),
        "new": len(cur - prior),
        "carried": len(cur & prior),
    }


def latest_prior_raw(history_dir):
    """直近の prior audit-*.json の raw_findings。存在しなければ None。

    ファイル名でソートする。名前に UTC timestamp が入るため辞書順が時系列に
    なる。読めない、あるいは raw_findings を欠く prior ファイルは run を
    中断せずスキップする。
    """
    priors = sorted(glob.glob(str(history_dir / "audit-*.json")), reverse=True)
    for path in priors:
        try:
            with open(path) as fh:
                data = json.load(fh)
        except (OSError, ValueError):
            continue
        raw = data.get("raw_findings")
        if isinstance(raw, list):
            return raw
    return None


def git_branch():
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        branch = out.stdout.strip()
        return branch or "unknown"
    except (OSError, subprocess.SubprocessError):
        return "unknown"


def build_record(payload, branch, generated_at, delta):
    record = dict(payload)
    record["branch"] = branch
    record["generated_at"] = generated_at
    record["delta"] = delta
    return record


def main():
    raw_stdin = sys.stdin.read()
    try:
        payload = json.loads(raw_stdin)
    except ValueError as exc:
        fail(f"unparseable payload: {exc}")
    if not isinstance(payload, dict):
        fail("payload must be a JSON object")

    HISTORY_DIR.mkdir(parents=True, exist_ok=True)
    now = datetime.now(timezone.utc)

    prior_raw = latest_prior_raw(HISTORY_DIR)
    current_raw = payload.get("raw_findings") or []
    delta = compute_delta(current_raw, prior_raw)

    record = build_record(
        payload,
        branch=git_branch(),
        generated_at=now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        delta=delta,
    )

    out_path = HISTORY_DIR / f"audit-{now.strftime('%Y-%m-%d-%H%M%S')}.json"
    with open(out_path, "w") as fh:
        json.dump(record, fh, ensure_ascii=False, indent=2)
    print(str(out_path))


if __name__ == "__main__":
    main()

File: .ja/workflows/audit/snapshot.ts
#!/usr/bin/env node
/// <reference types="node" />
// Usage: snapshot.ts   (audit payload JSON on stdin)
//
// audit の 1 実行を $HOME/.claude/history/audit-<YYYY-MM-DD-HHMMSS>.json に記録する。
//
// stdin:  JSON {scope, focus, pre_flight, raw_findings[], findings[], skipped[],
//         challenge_ran, verify_ran, tally, needs_context[], zero_reviewer_files[]}
//         各 raw_findings entry は最低限 {file, message} を持ち、triage を通った後は
//         {id, reviewer, verdict} も持つ。
//         キーは record にそのまま写す。無いキーは無いまま。
// stdout: JSON 1 行、{path, counts}。counts はこのプロセスが serialize した各配列の
//         要素数を持つ。
// exit 0 は成功。exit 1 は payload が parse 不能 (何も書かない)。
//
// record に追加される解決済みフィールド:
//   branch        git rev-parse --abbrev-ref HEAD ("unknown" にフォールバック)
//   generated_at  UTC ISO-8601
//
// 置き換え元の Python 版 audit recorder の TypeScript 移植。
// Contract: この CLI 自身の挙動。workflows/audit/tests/snapshot.test.ts が、固定 fixture
// workflows/audit/tests/fixtures/snapshot-cases.json に対してエンドツーエンドで検査する。
import { readFileSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { homedir } from "node:os";
import { historyPath, isoTimestamp, parsePayload } from "../_lib/cli.ts";
import { isMainModule } from "../_lib/entry-point.ts";

export const COUNTED_ARRAYS = [
  "raw_findings",
  "findings",
  "skipped",
  "needs_context",
  "zero_reviewer_files",
] as const;

/** `git rev-parse --abbrev-ref HEAD`。非 0 終了時や spawn 自体のエラー (PATH に git が無い)
 * では "unknown" にフォールバックする。Python 版の git_branch を写す。
 * Python 版の subprocess.run(timeout=10) と同じく 10s の timeout で走らせる。 */
export function gitBranch(): string {
  const result = spawnSync("git", ["rev-parse", "--abbrev-ref", "HEAD"], {
    encoding: "utf8",
    timeout: 10_000,
  });
  if (result.error || result.status !== 0) return "unknown";
  const branch = result.stdout.trim();
  return branch || "unknown";
}

/** 無いキーは省かず 0 と数える。呼び出し元が毎回同じキー集合を読めるので、配列が丸ごと
 * 落ちた場合もフィールドの欠落でなく件数の不一致として出る。Python 版の
 * counted_arrays を写す。 */
export function countedArrays(record: Record<string, unknown>): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const key of COUNTED_ARRAYS) {
    const value = record[key];
    counts[key] = Array.isArray(value) ? value.length : 0;
  }
  return counts;
}

/** disk に書く record: payload 自身のキーを元の順序のまま持ち、続けて `branch`、
 * `generated_at` を加える。Python 版の build_record を写す。 */
export function buildRecord(
  payload: Record<string, unknown>,
  branch: string,
  generatedAt: string,
): Record<string, unknown> {
  return { ...payload, branch, generated_at: generatedAt };
}

export function main(): number {
  const raw = readFileSync(0, "utf8");
  const { payload, message } = parsePayload(raw);
  if (payload === null) {
    process.stderr.write(`${message}\n`);
    return 1;
  }

  // 1 つの `now` が record の generated_at と record file 自身の timestamp 部分の両方に
  // 供給されるため、常に両者は一致する。Python 版の単一の `now` を写す。
  const now = new Date();
  const generatedAt = isoTimestamp(now);
  const stamp = generatedAt.replace(/:/g, "").replace("T", "-").replace("Z", "");

  const record = buildRecord(payload, gitBranch(), generatedAt);

  // historyPath は $HOME/.claude/history の存在を保証し full path を返す。
  // workflows/build/record.ts が自分の history file に使うのと同じ flow -- この recorder
  // との違いは `name` が固定の .jsonl 名ではなく run ごとに発行される点だけ。
  const outPath = historyPath(homedir(), `audit-${stamp}.json`);
  writeFileSync(outPath, `${JSON.stringify(record, null, 2)}\n`);

  const output = { path: outPath, counts: countedArrays(record) };
  process.stdout.write(`${JSON.stringify(output)}\n`);
  return 0;
}

if (isMainModule(import.meta.url)) {
  process.exit(main());
}
